<a href="https://colab.research.google.com/github/howellery/MPM_200_Howell/blob/main/Code/MPM200_Session9_HW_EHowell.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Python Packages and Scientific Computing Libraries

"""
Python Packages Overview:
-------------------------
Packages in Python are collections of modules that provide reusable code and functionality.
They allow developers to organize and share complex code libraries that can be easily imported
and used across different projects.

Key Characteristics of Packages:
- Organized hierarchically
- Contain multiple related modules
- Can be installed using package managers like pip
- Extend Python's core functionality

Three Essential Data Science Libraries:
--------------------------------------
1. Pandas (pd):
   - Primary library for data manipulation and analysis
   - Provides DataFrame: a 2D labeled data structure with columns of potentially different types
   - Key Features:
     * Reading/writing data from various sources (CSV, Excel, SQL databases)
     * Data cleaning and preprocessing
     * Complex data transformations
     * Powerful data analysis and aggregation tools

2. NumPy (np):
   - Fundamental package for scientific computing in Python
   - Introduces the ndarray (n-dimensional array) for efficient numerical operations
   - Key Features:
     * High-performance multidimensional array object
     * Mathematical and statistical functions
     * Linear algebra operations
     * Foundation for scientific and numerical computing
     * Enables fast computations on large datasets

3. Matplotlib (plt):
   - Comprehensive library for creating static, animated, and interactive visualizations
   - Provides MATLAB-like plotting interface
   - Key Features:
     * Line plots, scatter plots, bar charts, histograms
     * Customizable plot styles and layouts
     * Support for multiple plot types and graph formats
     * Integration with NumPy for data visualization

Practical Significance:
----------------------
These libraries form the backbone of data science and scientific computing in Python,
enabling complex data analysis, numerical computations, and insightful visualizations
with minimal code.
"""

In [462]:
# Import required libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import openpyxl as op
from IPython.display import Image, display

# Read the Excel file
# df = pd.read_excel("Flavivirus host associations.xlsx", )
url = "https://raw.github.com/howellery/MPM_200_Howell/refs/heads/main/Data/Flavivirus%20host%20associations.xlsx"
df = pd.read_excel(url, engine='openpyxl')

# Display basic information about the dataset
print(df.info())

# Show the first few rows of the data
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 570 entries, 0 to 569
Data columns (total 10 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   sr no.                  570 non-null    int64  
 1   Virus                   570 non-null    object 
 2   Scientific Name         570 non-null    object 
 3   Authors                 570 non-null    object 
 4   Year                    430 non-null    object 
 5   Title/accession number  570 non-null    object 
 6   Journal                 425 non-null    object 
 7   Issue                   415 non-null    float64
 8   Volume                  404 non-null    object 
 9   Pages                   404 non-null    object 
dtypes: float64(1), int64(1), object(8)
memory usage: 44.7+ KB
None


sr no.         Virus     Scientific Name  \
0       3    Apoi virus       Rattus rattus   
1       2    Apoi virus    Myodes rufocanus   
2       0    Apoi virus  Apodemus argenteus   
3       1    Apoi virus  Apodemus speciosus   
4       4  Bagaza virus      Alectoris rufa   

                                             Authors  Year  \
0  C. M. Baak-Baak, N. Cigarroa-Toledo, O. M. Tor...  2016   
1                                        CDC Arbocat   NaN   
2                 J. S. Mackenzie and D. T. Williams  2009   
3                 J. S. Mackenzie and D. T. Williams  2009   
4  D. Buitrago, A. Rocha, C. Tena-Tomas, M. Vigo,...  2012   

                              Title/accession number  \
0  New Records of Ectoparasites Echinolaelaps ech...   
1                                        CDC Arbocat   
2  The zoonotic flaviviruses of southern, south-e...   
3  The zoonotic flaviviruses of southern, south-e...   
4  Real-time fluorogenic reverse transcription po...   

                     Journal  Issue Volume   Pages  
0  Southwestern Entomologist    NaN    NaN     NaN  
1                        NaN    NaN    NaN     NaN  
2     Zoonoses Public Health   56.0    6-7  338-56  
3     Zoonoses Public Health   56.0    6-7  338-56  
4         J Vet Diagn Invest   24.0      5  959-63

## Using pandas functions that you have seen in the class address following questions
1. How many entries are there in the dataset?
2. How many viruses are present in the dataset?
3. List those viruses.
4. How many rows of data are present for each virus?
5. How many host species are present in the data?
6. Make a list of host species present.
7. Count the number of unique host species for each virus.
8. Which virus has the most number of hosts?
9. Which host has most number of viruses?

In [452]:
# Exploring Virus and Species Diversity in the Dataset

df.columns

# Count total number of entries in the dataset

print(df.shape)
print("There are 570 entries (570 rows) in the dataset.\n")

# Count unique viruses

print(df['Virus'].nunique())
print("There are 30 viruses in the dataset.\n")

# List of unique viruses

print("The unique viruses are:\n")
print(df['Virus'].unique())

# Count of each unique virus

print("\nThe count of each virus is:")
print(df.groupby('Virus').agg({'sr no.':'size'}))

# Count unique host species

print("\nThe number of unique host species is:")
print(df['Scientific Name'].nunique())

# List of unique host species

print("\nThe list of unique host species is:\n")
print(df['Scientific Name'].unique())

# Count of host species for each virus

print("\nThe count of host species for each virus is:")
countsdf = df.groupby('Virus').agg({'Scientific Name': 'nunique'})
countsdf = countsdf.rename(columns={'Scientific Name': 'Host_Count'})
#countsdf = df.groupby('Virus')['Scientific Name'].nunique()
print(countsdf)

# Most diverse viruses (viruses with most unique host species)

print("\nThe most diverse viruses, or viruses with the most unique host species are:")
countsdf = countsdf.sort_values(by='Host_Count', ascending=False)
print(countsdf[0:5])




(570, 10)
There are 570 entries (570 rows) in the dataset.

30
There are 30 viruses in the dataset.

The unique viruses are:

['Apoi virus' 'Bagaza virus' 'Banzi virus' 'Bouboui virus'
 'Bussuquara virus' 'Cacipacore virus' 'Dakar bat virus' 'Dengue virus'
 'Entebbe bat virus' 'Iguape virus' 'Ilheus virus'
 'Japanese encephalitis virus' 'Koutango virus' 'Kunjin virus'
 'Kyasanur forest virus' 'Louping-ill virus' 'Modoc virus'
 'Murray Valley encephalitis virus' 'Omsk hemorrhagic fever virus'
 'Rio Bravo virus' 'Rocio virus' 'St. Louis Encephalitis virus'
 'Tembusu virus' 'Tickborne encephalitis virus' 'Uganda S virus'
 'Usutu virus' 'Wesselsbron virus' 'West Nile virus' 'Yellow fever virus'
 'Zika virus']

The count of each virus is:
                                  sr no.
Virus                                   
Apoi virus                             4
Bagaza virus                           2
Banzi virus                            2
Bouboui virus                          1
Bussuquara

# Metadata

## Given that we are interested in the metadata associated with these host species, we will import datasets associated with the metadata for both bird and mammalian species.

1. Birds

In [453]:
loc = url = "https://raw.github.com/howellery/MPM_200_Howell/refs/heads/main/Data/Handbook%20of%20the%20Birds%20of%20the%20World%20and%20BirdLife%20International%20Digital%20Checklist%20of%20the%20Birds%20of%20the%20World_Version_9.xlsx"
birds_meta = pd.read_excel(loc, engine='openpyxl')
birds_meta.head()

Seq.  Subsp. Seq.             Order    Family name     Family Subfamily  \
0   1.0          0.0  STRUTHIONIFORMES  Struthionidae  Ostriches       NaN   
1   1.0          1.0               NaN            NaN        NaN       NaN   
2   1.0          2.0               NaN            NaN        NaN       NaN   
3   1.0          3.0               NaN            NaN        NaN       NaN   
4   1.0          4.0               NaN            NaN        NaN       NaN   

  Tribe     Common name             Scientific name       Authority  \
0   NaN  Common Ostrich            Struthio camelus  Linnaeus, 1758   
1   NaN             NaN   Struthio camelus syriacus             NaN   
2   NaN             NaN    Struthio camelus camelus             NaN   
3   NaN             NaN  Struthio camelus massaicus             NaN   
4   NaN             NaN  Struthio camelus australis             NaN   

  2024 IUCN Red List category Synonyms Alternative common names  \
0                          LC      NaN                      NaN   
1                         NaN      NaN                      NaN   
2                         NaN      NaN                      NaN   
3                         NaN      NaN                      NaN   
4                         NaN      NaN                      NaN   

                                 Taxonomic source(s)    SISRecID  \
0  del Hoyo, J., Collar, N.J., Christie, D.A., El...  45020636.0   
1                                                NaN         NaN   
2                                                NaN         NaN   
3                                                NaN         NaN   
4                                                NaN         NaN   

   SpcRecID\n(used before SISRecID)    SubsppID  
0                         1016860.0         NaN  
1                               NaN  45020636_1  
2                               NaN  45020636_2  
3                               NaN  45020636_3  
4                               NaN  45020636_4

The data for birds is at subspecies level. It means that there is species name represented by column `Seq.`. All rows in same number represent one species.

2. Mammals

## Similarly, read file for mammals metadata.

In [454]:
loc = "https://raw.github.com/howellery/MPM_200_Howell/refs/heads/main/Data/IUCN%20Mammals,%20Birds,%20Reptiles,%20and%20Amphibians.csv"
mammals_meta = pd.read_csv(loc, encoding='latin-1')
mammals_meta.head()

Species ID   Kingdom    Phylum     Class             Order       Family  \
0     42641.0  ANIMALIA  CHORDATA  MAMMALIA          RODENTIA      MURIDAE   
1  22687170.0  ANIMALIA  CHORDATA      AVES  CAPRIMULGIFORMES  TROCHILIDAE   
2     17879.0  ANIMALIA  CHORDATA  MAMMALIA          RODENTIA      MURIDAE   
3  47760825.0  ANIMALIA  CHORDATA  MAMMALIA          RODENTIA   CRICETIDAE   
4     42656.0  ANIMALIA  CHORDATA  MAMMALIA          RODENTIA  ABROCOMIDAE   

          Genus    Species                  Authority  Infraspecific rank  \
0     Abditomys   latidens            (Sanborn, 1952)                 NaN   
1      Abeillia   abeillei  (Lesson & DeLattre, 1839)                 NaN   
2   Abeomelomys      sevia    (Tate & Archbold, 1935)                 NaN   
3  Abrawayaomys    ruschii         Cunha & Cruz, 1979                 NaN   
4      Abrocoma  bennettii           Waterhouse, 1837                 NaN   

   ...                             Synonyms  \
0  ...                                  NaN   
1  ...                                  NaN   
2  ...  Melomys sevia|Pogonomelomys sevia     
3  ...                                  NaN   
4  ...                                  NaN   

                                  Common names (Eng)  Common names (Fre)  \
0                            Luzon Broad-toothed Rat                 NaN   
1                        Emerald-chinned Hummingbird                 NaN   
2  Menzies' Mouse, Menzies's Mouse, Highland Brus...                 NaN   
3                                       Ruschi's Rat                 NaN   
4                           Bennett's Chinchilla Rat                 NaN   

  Common names (Spa) Red List status Red List criteria  \
0                NaN              DD               NaN   
1                NaN              LC               NaN   
2                NaN              LC               NaN   
3                NaN              LC               NaN   
4                NaN              LC               NaN   

  Red List criteria version Year assessed Population trend  Petitioned  
0                       3.1        2008.0          unknown           N  
1                       3.1        2012.0       decreasing           N  
2                       3.1        2016.0          unknown           N  
3                       3.1        2016.0          unknown           N  
4                       3.1        2016.0          unknown           N  

[5 rows x 23 columns]

## For both birds and mammals answer the following questions.
1. Size of the datasets
2. For birds, filter the dataset using species-level rows only. Hint. Column `Subsp. Seq.` value of 0 indicates the main species name
3. How many bird species and mammal species are present in data frames?

In [455]:
# Size of the datasets:
print("The size of the animal datasets:\n")

print(birds_meta.shape)
print(mammals_meta.shape)

print("\nThere are 23 columns and 16081 rows in the mammal, birds, and reptiles set, and 17 columns and 33265 rows in the bird set.")

# Filter bird set to species level rows only:
birds_filter = birds_meta[birds_meta["Subsp. Seq."] == 0.0]
birds_filter = birds_filter.dropna(subset=['Scientific name'])
birds_filter.head()

# How many bird species and mammal species are present in the data frames?
print("\nThe number of unique birds in the Handbook of the Birds of the World and BirdLife International Digital Checklist \nof the Birds dataset is 11198.\n")

print(birds_filter['Scientific name'].nunique())

# Create a column in the mammal dataset for Scientific name to match the column in the bird dataset
# Also create a new mammals dataset with the Scientific Name column:
mammals_science = mammals_meta
mammals_science['Scientific name'] = mammals_meta['Genus'] + " " + mammals_meta['Species']
mammals_science = mammals_science.dropna(subset=['Scientific name'])

# Determine the number of unique birds, mammals, reptiles, and total animals
print("\nThe number of unique birds and mammals in the IUCN Mammals, Birds, Reptiles, and Amphibians dataset:\n")
mammals_science['Class'].unique()
print(mammals_science.groupby('Class').agg({'Scientific name':'nunique'}))
print('\nTotal unique animals:')
print(mammals_science['Scientific name'].nunique())

# Diagnose cause of discrepancy between table row number and number of unique animals
print("\nSince the number of unique animals in the birds, mammals, reptiles dataset don't add up to the total rows, there must be some duplicate animal names.")
print('Number of duplicate animals:')
print(mammals_science['Scientific name'].duplicated().sum())
print(mammals_science['Scientific name'].value_counts().head())

print('\nCombined number of unique animals in the datasets (or maximum number of unique animals when datasets are combined):', 11198+16073)


The size of the animal datasets:

(33265, 17)
(16081, 23)

There are 23 columns and 16081 rows in the mammal, birds, and reptiles set, and 17 columns and 33265 rows in the bird set.

The number of unique birds in the Handbook of the Birds of the World and BirdLife International Digital Checklist 
of the Birds dataset is 11198.

11198

The number of unique birds and mammals in the IUCN Mammals, Birds, Reptiles, and Amphibians dataset:

          Scientific name
Class                    
AVES                10452
MAMMALIA             5621

Total unique animals:
16073

Since the number of unique animals in the birds, mammals, reptiles dataset don't add up to the total rows, there must be some duplicate animal names.
Number of duplicate animals:
4
Scientific name
Marmosa demerarae     2
Murina aurata         2
Glis glis             2
Erithacus rubecula    2
Perdix perdix         1
Name: count, dtype: int64

Combined number of unique animals in the datasets (or maximum number of unique anim

# Database Integration Exercise
The main objective of this exercise is to create a comprehensive database by combining three data frames. Follow these steps  n your analysis.

## Create a concept map:  
* Illustrate the relationships between the three data frames.  
* Identify and highlight the Primary Keys in each data frame.  
* Use a diagramming tool (e.g., PowerPoint, LucidChart, or Draw.io) to create a visual representation of these connections.  
* Pay special attention to how you plan to merge the Mammals and Birds data frames.  

In [464]:
# Create a concept map
picture = "https://raw.github.com/howellery/MPM_200_Howell/refs/heads/main/Data/Data/MPM200Python.jpg"
display(Image(url=picture))


## Perform a vertical join of Mammals and Birds data frames:  
* Identify and select only the common columns present in both data frames.  
* Use `pd.concat()` or `df.append()` to create a larger dataframe called **metadata**.  
* Ensure the resulting dataframe has the correct number of rows and columns.  

In [457]:
# Perform a vertical join of Mammals and Birds data frames

print("\nThe number of duplicate rows in the two data sets is: \n")

# Number of overlapping rows in the data sets
overlap_count = birds_filter.loc[birds_filter['Scientific name'].isin(mammals_science['Scientific name']), 'Scientific name'].nunique()
print(overlap_count)

birds_filter['Family name'] = birds_filter['Family name'].str.upper()
print("\nBirds\n")
print(birds_filter.groupby('Family name').agg({'Scientific name':'nunique'}))
print("\nBirds & Mammals\n")
print(mammals_science.groupby('Family').agg({'Scientific name':'nunique'}))

# Find the common columns in the data sets:
print("\nThe columns in the bird data set is: \n")
print(birds_filter.columns.tolist())
print("\nThe columns in the birds and mammals data set is: \n")
print(mammals_science.columns.tolist())

print("\nThe common columns in the two data sets are: \n")
print(set(mammals_science.columns).intersection(birds_filter.columns))

print("\nOther columns with similar information: Family name in birds/Family in mammals datasets and Common name/Common names (Eng).\n")

birds_filter = birds_filter.rename(columns={'Family': 'Animal Type'})
birds_filter = birds_filter.rename(columns={'Family name': 'Family'})

mammals_science = mammals_science.rename(columns={'Common names (Eng)': 'Common name'})
print("\nIdeally I would combine on Common name column as well as the others, but since that column is likely to have a lot of variability I decided not to.\n")
print("\nIdeally I would also figure out how to match based on either Scientific name or Synonym columns.\n")

# Only select based on common columns:
BirdSubset = birds_filter[["Authority", "Scientific name", "Family", "Synonyms", "Order"]]
MammalSubset = mammals_science[["Authority", "Scientific name", "Family", "Synonyms", "Order"]]

print("\nThe combined data set: \n")
# Vertical join of the data sets
metadata = pd.concat([MammalSubset, BirdSubset], axis=0)

print(metadata.groupby('Family').agg({'Scientific name':'nunique'}))

print(metadata.shape)
print(metadata['Scientific name'].nunique())
metadata.head()

print(metadata['Scientific name'].duplicated().sum())

# New data set based only on Scientific name:

betterdata = pd.merge(MammalSubset, BirdSubset, on=["Scientific name", "Family", "Order"], how='outer')
print("Shape of data table with fewer duplicates:", betterdata.shape)
print(betterdata['Scientific name'].nunique())




The number of duplicate rows in the two data sets is: 

8916

Birds

                 Scientific name
Family name                     
ACANTHISITTIDAE                4
ACANTHIZIDAE                  65
ACCIPITRIDAE                 244
ACROCEPHALIDAE                60
AEGITHALIDAE                  14
...                          ...
VANGIDAE                      40
VIDUIDAE                      20
VIREONIDAE                    66
ZELEDONIIDAE                   1
ZOSTEROPIDAE                 144

[247 rows x 1 columns]

Birds & Mammals

                 Scientific name
Family                          
ABROCOMIDAE                   10
ACANTHISITTIDAE                4
ACANTHIZIDAE                  65
ACCIPITRIDAE                 250
ACROBATIDAE                    2
...                          ...
VIREONIDAE                    52
VIVERRIDAE                    33
VOMBATIDAE                     3
ZIPHIIDAE                     21
ZOSTEROPIDAE                 100

[363 rows x 1 columns]

The c

The number of columns that are in the new data frame is correct (5).

It's hard to tell if the number of rows that are include in the new data frame is correct. Since there are probably many overlapping species between the two, I can only tell that the new total unique number (18355) appears as less than the max possible combined number that I found before (27271). I decided to further check the matchups by ensuring that some of the combined groupings on Family add up to greater totals than the maximum of the two groupings on family between the two datasets. For example, there are 144 ZOSTEROPIDAE in the birds data set and 100 ZOSTEROPIDAE in the birds and mammals data set.

Another issue is that there could be scientific names that didn't match up properly due to the Synonyms column for one dataset containing the Scientific name listed in the other and vice versa. I could use a loop to search through every synonym for each Scientific name in order to check this, but that would take a really long time to run.

Since running code to show duplicates in the data shows that there are 8926 rows with duplicated Scientific names, I will re-combine the datasets with merge and call it "betterdata". The betterdata set is merged only based on Scientfic names, Family, and Order. The total number of rows is 19253, which means that there are only about 898 duplicates remaining (8926 - (27281 - 19253) = 898).


## Merge the metadata dataframe with the Flavivirus dataframe:  
* Experiment with different join types: `inner`, `outer`, `left`, and `right`.  
* For each join type, record the resulting dataframe's shape (rows and columns).  
* Determine the **theoretically correct size of the merged dataframe if done manually**.  
* Compare the results of each join type with the expected outcome.  
* Choose the most appropriate join type based on your analysis.  

In [458]:
# Merge the metadata dataframe with the Flavivirus dataframe

df = df.rename(columns={'Scientific Name': 'Scientific name'})

# Inner join (only rows that match exactly on the key)

Inner_Virus = pd.merge(df, metadata, on='Scientific name', how='inner')
print("Inner join table shape:", Inner_Virus.shape)

# Left join (all rows from the first table)

Left_Virus = pd.merge(df, metadata, on='Scientific name', how='left')
print("Left join table shape:", Left_Virus.shape)

# Right join (all rows from the second table)

Right_Virus = pd.merge(df, metadata, on='Scientific name', how='right')
print("Right join table shape:", Right_Virus.shape)

# Outer join (all rows in both tables even if they don't match on the key)

Outer_Virus = pd.merge(df, metadata, on='Scientific name', how='outer')
print("Outer join table shape:", Outer_Virus.shape)

# Left join (all rows from the first table)

betterjoin = pd.merge(df, betterdata, on='Scientific name', how='left')
print("Left join the betterdata table shape:", betterjoin.shape)

# Inner join (only rows that match exactly on the key)

betterjoininner = pd.merge(df, betterdata, on='Scientific name', how='inner')
print("Inner join the betterdata table shape:", betterjoininner.shape)

# Right join (all rows from the second table

betterjoinright = pd.merge(df, betterdata, on='Scientific name', how='right')
print("Right join the betterdata table shape:", betterjoinright.shape)

# Outer join (all rows in both tables even if they don't match on the key)

betterjoinouter = pd.merge(df, betterdata, on='Scientific name', how='outer')
print("Outer join the betterdata table shape:", betterjoinouter.shape)


Inner join table shape: (872, 14)
Left join table shape: (881, 14)
Right join table shape: (27501, 14)
Outer join table shape: (27510, 14)
Left join the betterdata table shape: (601, 16)
Inner join the betterdata table shape: (592, 16)
Right join the betterdata table shape: (19409, 16)
Outer join the betterdata table shape: (19418, 16)


The inner join shows that there are 872 of the rows match up on Scientific name, while the left join shows that there are 881 rows in the Flavivirus dataset that should match up.

Theoretically, the data set should include only 570 rows and 14 columns if done manually, since there were only 570 virus rows in the original dataset. However, since there are a few species duplicates in the animals dataset, the joins still produce more rows than 570. There are the appropriate number of columns in all of the joins. In the inner join, the 872 rows indicate that compared to the left join with 881 rows, there are 9 species that don't match up perfectly between the two sets. The same is indicated by the difference of 9 columns between the outer join and the right join. I care the most about retaining all the virus data, so using the left join is most appropriate.

To fix the problem overall, I minimized the number of duplicates with the betterdata version of the animal datasets. When left joined to the Flavirus dataset, the betterdata merge produces 601 rows, which is only 31 rows greater than ideal. There are also 2 duplicate columns due to the way betterdata was created, but overall this dataset is closer to correct.
   
  


### Create a pivote table,
1. Rows are viruses  
2. Columns are Taxonomic orders  
3. Values are number of unique host species  

In [459]:
# Create a pivote table

betterjoin.groupby('Virus').agg({'Family':'nunique','Order':'nunique', 'Scientific name':'nunique'})


Family  Order  Scientific name
Virus                                                           
Apoi virus                             2      1                4
Bagaza virus                           1      1                2
Banzi virus                            2      1                2
Bouboui virus                          1      1                1
Bussuquara virus                       3      3                3
Cacipacore virus                       1      1                1
Dakar bat virus                        3      1                5
Dengue virus                          11      5               27
Entebbe bat virus                      4      1                6
Iguape virus                           1      1                1
Ilheus virus                           5      3                5
Japanese encephalitis virus           14      7               25
Koutango virus                         1      1                1
Kunjin virus                           3      3                3
Kyasanur forest virus                  3      3                6
Louping-ill virus                      9      7               13
Modoc virus                            2      1                3
Murray Valley encephalitis virus       2      2                3
Omsk hemorrhagic fever virus           1      1                3
Rio Bravo virus                        3      1                9
Rocio virus                            3      2                2
St. Louis Encephalitis virus          42     19               72
Tembusu virus                          1      1                2
Tickborne encephalitis virus          34     13               83
Uganda S virus                         3      2                4
Usutu virus                           33     18               65
Wesselsbron virus                      3      3                3
West Nile virus                       71     26              194
Yellow fever virus                     5      2               11
Zika virus                             4      1                7

I know that this table is accurate because the number of each species (Scientific name column) is equal to what I found for species number by virus in the first exercise with the Flavivirus set.